model/task_queue.go

```diff
@@ -74,7 +79,57 @@ func (self *TaskQueue) Length() int {
 }
 
 func (self *TaskQueue) NextTask() *TaskQueueItem {
-	return &self.Queue[0]
+	for _, it := range self.Queue {
+		// Always return a task if the task group is empty.
+		if it.Group == "" {
+			return &it
+		}
+		// Otherwise, return the task if it is running on fewer than its task group's max hosts.
+		spec := TaskSpec{
+			Group:         it.Group,
+			BuildVariant:  it.BuildVariant,
+			ProjectID:     it.Project,
+			Version:       it.Version,
+			GroupMaxHosts: it.GroupMaxHosts,
+		}
+		if shouldRun := shouldRunTaskGroup(it.Id, spec); shouldRun {
+			return &it
+		}
+	}
+	return nil
+}
```

Comment:
I think this should go in the FindTask implementation not NextTask.

Reply:
done